src/torrent_info.cpp

```diff
@@ -640,58 +644,117 @@ namespace {
+			boost::crc_optimal<32, 0x1EDC6F41, 0xFFFFFFFF, 0xFFFFFFFF, true, true> crc;
+			if (!m_files.name().empty())
 			{
-				p = parent_path(std::move(p));
-				// we don't want trailing slashes here
-				TORRENT_ASSERT(p[p.size() - 1] == TORRENT_SEPARATOR);
-				p.resize(p.size() - 1);
-				files.insert({p, file_index_t{-1}});
+				process_string_lowercase(crc, m_files.name());
+			}
+			file_index_t path_index{-1};
+			for (auto const& path : paths)
+			{
+				auto local_crc = crc;
+				if (!path.empty()) local_crc.process_byte(TORRENT_SEPARATOR);
+				int count = 0;
+				for (char const c : path)
+				{
+					if (c == TORRENT_SEPARATOR)
+						files.insert({local_crc.checksum(), {path_index, count}});
+					local_crc.process_byte(to_lower(c) & 0xff);
+					++count;
+				}
+				files.insert({local_crc.checksum(), {path_index, int(path.size())}});
+				--path_index;
 			}
 		}
 
+		// keep track of the total number of name collisions. If there are too
+		// many, it's probably a malicious torrent and we should just fail
+		int num_collisions = 0;
 		for (auto const i : m_files.file_range())
 		{
 			// as long as this file already exists
 			// increase the counter
-			std::string filename = m_files.file_path(i);
-			auto const ret = files.insert({filename, i});
-			if (ret.second) continue;
+			std::uint32_t const hash = m_files.file_path_hash(i, "");
+			auto range = files.equal_range(hash);
+			auto const match = std::find_if(range.first, range.second, [&](std::pair<std::uint32_t, name_entry> const& o)
+			{
+				std::string const other_name = o.second.idx < file_index_t{}
+					? combine_path(m_files.name(), paths[std::size_t(-static_cast<int>(o.second.idx)-1)].substr(0, std::size_t(o.second.length)))
+					: m_files.file_path(o.second.idx);
+				return string_equal_no_case(other_name, m_files.file_path(i));
+			});
+
+			if (match == range.second)
+			{
+				files.insert({hash, {i, 0}});
+				continue;
+			}
+
 			// pad files are allowed to collide with each-other, as long as they have
 			// the same size.
-			file_index_t const other_idx = ret.first->second;
-			if (other_idx != file_index_t{-1}
+			file_index_t const other_idx = match->second.idx;
+			if (other_idx >= file_index_t{}
 				&& (m_files.file_flags(i) & file_storage::flag_pad_file)
 				&& (m_files.file_flags(other_idx) & file_storage::flag_pad_file)
 				&& m_files.file_size(i) == m_files.file_size(other_idx))
 				continue;
 
+			std::string filename = m_files.file_path(i);
 			std::string base = remove_extension(filename);
 			std::string ext = extension(filename);
 			int cnt = 0;
-			do
+			for (;;)
 			{
 				++cnt;
 				char new_ext[50];
 				std::snprintf(new_ext, sizeof(new_ext), ".%d%s", cnt, ext.c_str());
 				filename = base + new_ext;
+
+				boost::crc_optimal<32, 0x1EDC6F41, 0xFFFFFFFF, 0xFFFFFFFF, true, true> crc;
+				process_string_lowercase(crc, filename);
+				std::uint32_t const new_hash = crc.checksum();
+				if (files.find(new_hash) == files.end())
```

Comment:
I'm not super happy about this one. it's a bit sloppy to *just* check for a hash collision, and not also check whether the name collides. The idea is that it doesn't really matter that much if someone can construct names that will collide, since the worst case is that the sequence numbers appended to the filenames aren't contiguous.

and I'm pretty sure you can't trigger an infinite loop here, CRC32 is really good at covering all the bits, so you would need near `UINT32_MAX` file entries for that